plug.vim

```diff
@@ -1980,17 +1987,24 @@ function! s:update_ruby()
 EOF
 endfunction
 
-function! s:shellesc_cmd(arg)
-  let escaped = substitute(a:arg, '[&|<>()@^]', '^&', 'g')
-  let escaped = substitute(escaped, '%', '%%', 'g')
-  let escaped = substitute(escaped, '"', '\\^&', 'g')
-  let escaped = substitute(escaped, '\(\\\+\)\(\\^\)', '\1\1\2', 'g')
-  return '^"'.substitute(escaped, '\(\\\+\)$', '\1\1', '').'^"'
+function! s:shellesc_cmd(arg, script)
+  let escaped = substitute(a:arg, '%', (a:script ? '%' : '^') . '&', 'g')
+  return substitute('"'.escaped.'"', '[&|<>()@^!"]', '^&', 'g')
 endfunction
 
-function! s:shellesc(arg)
-  if &shell =~# 'cmd.exe$'
-    return s:shellesc_cmd(a:arg)
+function! s:shellesc_ps1(arg)
+  return "'".substitute(escape(a:arg, '\"'), "'", "''", 'g')."'"
+endfunction
+
+function! s:shellesc(arg, ...)
```

Comment:
It's a little hard for me to see what to expect from this function especially with the extra arguments. It would be great if we have functional test cases for `s:shellesc`. There are a few issues though.

- Vader tests are written for Unix environment. They probably won't work correctly on Windows.
- `s:shellesc` is script-local so it's not possible to call it directly.
    - Related, we are testing `s:compare_git_uri` by patching `plug.vim` file in a very hacky way
        - https://github.com/junegunn/vim-plug/blob/226d6abeb2f02bcd8f5b11288543259fa2e02962/test/test.vader#L15-L19
        - https://github.com/junegunn/vim-plug/blob/226d6abeb2f02bcd8f5b11288543259fa2e02962/test/workflow.vader#L1633-L1658
    - But the patching code may not work on Windows, and it's not easy to run the test cases separately

So what I'm suggesting is that

1. We make it visible for tests:

    We could do it conditionally only when some environment variable is set,
    ```vim
    " At the end of plug.vim
    if len($PLUG_TEST)
      function! Shellesc(...)
        return call('s:shellesc', a:000)
      endfunction
    endif
    ```
    or just expose it as `plug#shellescape` and keep it undocumented. I'm fine with either way.

2. And add a new Vader file (`functional.vader`) we can easily run.